---
title: "CQS Discovery Oritented Data Science - scRNA-seq data analysis"
output: html_notebook
---

# Intro
##### Author: Paige (Spencer) Vega, Vanderbilt University, Ken Lau lab
##### This tutorial was modified from Seurat's tutorials available at the links below:
##### https://satijalab.org/seurat/pbmc3k_tutorial.html
##### https://satijalab.org/seurat/immune_alignment.html
##### You can also learn more about the Seurat tool by reading their paper:
##### https://www.nature.com/articles/nbt.4096


# Data information
##### Kang et al., 2017 (https://www.nature.com/articles/nbt.4042)
##### to download it, follow this link:
##### https://www.dropbox.com/s/79q6dttg8yl20zg/immune_alignment_expression_matrices.zip?dl=1
##### The first dataset is 13,019 PBMCs (peripheral blood mononuclear cells).  The second is 12,875 PBMCs stimulated with interferon gamma.


# Learning goals for section 1
#####   1.  Check the quality of the data and filter it.
#####   2.  Normalize, scale, and find highly variable genes (HVGs).
#####  3.  Dimension reduction.
#####   4.  Differential gene expression.




# Call on the Seurat function

```{r}
# load in the Seurat package
library(Seurat)

# set your working directory
directory<-setwd("/Users/msenosain/Documents/GitHub_Mafe/Discovery_Oriented_Data_Science")

```


# Load in your data and create Seurat object

```{r}
# load in the data
  # 35,635 rows (genes)
  # 13,019 cells (names are barcodes)

pbmc.data <- read.table(paste(directory,"/data/immune_control_expression_matrix.txt.gz", sep = ""), sep = "\t")

# Create the Seurat object with the raw (non-normalized data).  Keep all genes expressed in >= 5 cells. 

pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 5, project = "control_PBMC")

# pbmc@raw.data

```


# Filtering

```{r}
# The number of genes and UMIs (nGene and nUMI) are automatically calculated
# for every object by Seurat. Use violin plots to visualize number genes and number UMIs.

VlnPlot(object = pbmc, features.plot = c("nGene", "nUMI"), nCol = 2, point.size.use = 0.01)

# Note: finding the percent mitochondrial genes for a dataset is a common QC metric, but, this dataset is already pre-processed to remove cells that have mitochondrial gene expression.
```

```{r}
# GenePlot is typically used to visualize gene-gene relationships, but can be used to see how any two variables correlate.  We will use it to see how number UMIs correlate to number of genes.

# by the way, what would be "bad" and what would be "good"?

      # bad = low nGene with high nUMI => means there are few types of genes with high expression
      # good = linear correlation nGene to nUMI, with nUMI about 3-5X greater than nGene

GenePlot(object = pbmc, gene1 = "nUMI", gene2 = "nGene")

# Note: if you're trying to remove cells expressing high % mitochondrial genes, plot nUMI to % mito and QC from there.
```

```{r}
# Based on the violin plot, filter out cells by choosing low and high thresholds.

# get rid of cells with very few nGenes or way too many nGene
pbmc <- FilterCells(object = pbmc, subset.names = "nGene", low.thresholds = 200, high.thresholds = 1800)

# if you were removing cells with high % mito expression, remove them here with appropriate thresholds.

# pbmc@raw.data
```



# Normalization

```{r}
# Normalize the data - log normalization is default.

pbmc <- NormalizeData(object = pbmc)

# pbmc@data
```


# Find highly variable genes (HVGs)

```{r}
# Find the top 1,000 most variable genes
  # Calculates average expression and dispersion, using z-scores to determine outliers, which        accounts for the relationship between variability and avg expression.

pbmc <- FindVariableGenes(pbmc)

# pbmc@var.genes # contains variable genes
# pbmc@hvg.info  # contains results of HVG analysis
```

# Scale data

```{r}
# removes "uninteresting sources of variation", like technical noise, or cell cycle
# linear regression to predict gene expression. See ?ScaleData for more information.

pbmc <- ScaleData(pbmc)

# pbmc@scale.data
```

# Dimensionality reduction (PCA)

```{r}
# Run PCA using the HVGs

pbmc <- RunPCA(object = pbmc, pc.genes = pbmc@var.genes, pcs.compute = 30, do.print = FALSE)

# pbmc@dr$pca

```

# Choose PCs

```{r fig.height=6, fig.width=10}
# Principle components (PCs) capture the variability in your dataset.  Although we specified 30 PCs, we want to use only the PCs that capture variability and discard the rest for downstream analyses.  You look for saturation (flatline) in the relationship between the number of principle components and the percentage of the variance explained.

# 3 ways to look at PCs...

# 1. Elbow plot of principal components
PCElbowPlot(object = pbmc, num.pc = 30)

# 2. Heatmap of individual PCs
PCHeatmap(object = pbmc, pc.use = 1:15, cells.use = 500, do.balanced = TRUE, label.column = FALSE, use.full = FALSE)

PCHeatmap(object = pbmc, pc.use = 16:30, cells.use = 500, do.balanced = TRUE, label.column = FALSE, use.full = FALSE)

# 3. Jackstraw package to determine statistically significant principal components
#pbmc <- JackStraw(object = pbmc, num.pc = 30, num.replicate = 100, display.progress = TRUE)
#JackStrawPlot(object = pbmc, PCs = 1:30)

```

# Use PCs to cluster and visualize clustering with t-SNE

```{r}
# Clustering using Seurat's method. See more info in ?FindClusters.
  # specify the number of PCs using dims.use

pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:20, resolution = 0.6,
                     print.output = 0, save.SNN = TRUE)
# pbmc@ident

# Run dimension reduction (t-SNE) to visualize the clustering
pbmc <- Seurat::RunTSNE(object = pbmc, dims.use = 1:20, do.fast = TRUE)

TSNEPlot(object = pbmc)

```


# Differential gene expression - one cluster at a time

```{r}
# cluster 1 vs all other clusters pooled together
cluster1.markers <- FindMarkers(object = pbmc, ident.1 = 1, min.pct = 0.25)
print(x = head(x = cluster1.markers, n =5))

# cluster 1 vs cluster 5
cluster1.markers.5 <- FindMarkers(object = pbmc, ident.1 = 1, ident.2 = 5, min.pct = 0.25)
print(x = head(x = cluster1.markers.5, n =5))


```

# Differential gene expression - iterate through all clusters

```{r}
# Find markers for every cluster compared to all remaining cells, report only positive ones
pbmc.markers <- FindAllMarkers(object = pbmc, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)


# What about using a different test?  On your own time, trying adding the argument test.use = "roc" and see how different your results are. 
```


# Visualize differentially expressed genes

```{r}
# The list of DE genes is long and it would be too much information to visualize them all. So, let's plot the top 5 features (genes) for each cluster.  

# call library dplyr to use some of its functions
library(dplyr)
# group by cluster using group_by function and show top 5 results using top_n function
pbmc.markers.grouped <- group_by(pbmc.markers, cluster)
pbmc.markers.grouped.top <- top_n(pbmc.markers.grouped, 5, avg_logFC)
# convert it to a matrix for downstream applications (trust me)
pbmc.markers.grouped.top <- as.matrix(pbmc.markers.grouped.top)
print(pbmc.markers.grouped.top)

# First, isolate the top 5 DE genes for each cluster.
cluster_0 <- which(pbmc.markers.grouped.top[,"cluster"] == 0)
cluster_1 <- which(pbmc.markers.grouped.top[,"cluster"] == 1)
cluster_2 <- which(pbmc.markers.grouped.top[,"cluster"] == 2)
cluster_3 <- which(pbmc.markers.grouped.top[,"cluster"] == 3)
cluster_4 <- which(pbmc.markers.grouped.top[,"cluster"] == 4)
cluster_5 <- which(pbmc.markers.grouped.top[,"cluster"] == 5)
cluster_6 <- which(pbmc.markers.grouped.top[,"cluster"] == 6)
cluster_7 <- which(pbmc.markers.grouped.top[,"cluster"] == 7)
cluster_8 <- which(pbmc.markers.grouped.top[,"cluster"] == 8)
cluster_9 <- which(pbmc.markers.grouped.top[,"cluster"] == 9)
cluster_10 <- which(pbmc.markers.grouped.top[,"cluster"] == 10)


# 3 options for visualization...

# 1. Violin plots
# if you want to see the points, insert point.size.use = 0.01 into the argument list

# # specify cluster 0 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_0, "gene"],
        point.size.use = 0)
# # specify cluster 1 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_1, "gene"],
        point.size.use = 0)
# # specify cluster 2 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_2, "gene"],
        point.size.use = 0)
# # specify cluster 3 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_3, "gene"],
        point.size.use = 0)
# # specify cluster 4 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_4, "gene"],
        point.size.use = 0)
# # specify cluster 5 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_5, "gene"],
        point.size.use = 0)
# # specify cluster 6 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_6, "gene"],
        point.size.use = 0)
# # specify cluster 7 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_7, "gene"],
        point.size.use = 0)
# # specify cluster 8 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_8, "gene"],
        point.size.use = 0)
# # specify cluster 9 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_9, "gene"],
        point.size.use = 0)
# # specify cluster 10 here
VlnPlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_10, "gene"],
        point.size.use = 0)



# 2. Feature plot
# You could change the reduction.use to do PCA or ICA instead of t-SNE, but since we know where our clusters were mapped in the context of t-SNE, this makes the most sense for us.

# Cluster 0
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_0, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 1
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_1, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 2
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_2, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 3
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_3, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 4
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_4, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 5
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_5, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 6
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_6, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 7
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_7, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 8
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_8, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 9
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_9, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")
# Cluster 10
FeaturePlot(object = pbmc, features.plot = pbmc.markers.grouped.top[cluster_10, "gene"], cols.use = c("grey", "blue"), reduction.use = "tsne")


# 3. Heatmap
# Heatmaps have enough room to show you the top 5 (or more!) genes for each cluster.  Let's do the top 5, but if you have time, try changing the top features to 10 or more.

DoHeatmap(object = pbmc, genes.use = pbmc.markers.grouped.top[ , "gene"], slim.col.label = TRUE, remove.key = TRUE)

```

# Save the object and workspace

```{r}
# save this object and come back to it later
saveRDS(pbmc, file = paste(directory, "/pbmc_only.rds", sep = ""))

# save work in R
save.image(paste(directory, "/section1.RData", sep = ""))

```

save.image(paste(directory, "/section1.RData", sep = ""))



# Now let's add in a second dataset, which is PBMCs stimulated with interferon gamma

```{r}
# load in the data - how many cells and how many genes?  use dim(stim.data) to find out
stim.data <- read.table(paste(directory, "/immune_stimulated_expression_matrix.txt.gz", sep = ""), sep = "\t")

# create the Seurat object for the stimulated PBMCs
stim <- CreateSeuratObject(raw.data = stim.data, min.cells = 5, project = "stim_PBMC")

# Also re-create the control PBMC object because we will change a couple things this time and we don't want to have the previous info stored in the Seurat object.
pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 5, project = "control_PBMC")

```

#### We have a few goals in this section of the analysis.  
#####   1.  identify cell types present in both the control pbmc and stimulated pbmc datasets
#####   2.  find markers that identify cell types that are conserved in control and stimulated pbmcs
#####   3.  compare the datasets to find cell-type specific responses to interferon stimulation 

#### Really, the goal is to reinforce the major steps in scRNA-seq analysis by going over some of them again.  This time, we show how even a simple experimental design (control vs stimulated) can complicate the analysis.



# Filtering, normalization, scaling, and finding HVGs
```{r}

VlnPlot(object = stim, features.plot = c("nGene", "nUMI"), nCol = 2, point.size.use = 0.001)

````

```{r}
GenePlot(object = stim, gene1 = "nUMI", gene2 = "nGene")
```

```{r}
# Filter out the stimulated pbmcs with low total gene expression
  # for this analysis, we want only the genes with at least average nGene
stim <- FilterCells(object = stim, subset.names = "nGene", low.thresholds = 500, high.thresholds = 1800)
# normalize
stim <- NormalizeData(stim)
# scale
stim <- ScaleData(stim)
# Find variable genes
stim <-FindVariableGenes(stim)
# make metadata indicating it is stimulated
stim@meta.data$stim <- "STIM"

# yes, we already did this for the control dataset, but the downstream analysis requires a bit more selective filtering for it to run.  The only changes we are making is choosing cells with 500-1800 genes instead of 200-1800 genes.
pbmc <- FilterCells(object = pbmc, subset.names = "nGene", low.thresholds = 500, high.thresholds = 1800)
# normalize
pbmc <- NormalizeData(pbmc)
# scale
pbmc <- ScaleData(pbmc)
# Find variable genes
pbmc <-FindVariableGenes(pbmc)
# make metadata indicating it is stimulated
pbmc@meta.data$stim <- "CTRL"


```

```{r}
# Now find the highly variable genes that are shared between the two datasets (control and stimulated)

g.1 <- head(rownames(pbmc@hvg.info), 1000)
g.2 <- head(rownames(stim@hvg.info), 1000)

# keeps only unique genenames (gets rid of duplicates)
genes.use <- unique(c(g.1, g.2))
genes.use <- intersect(genes.use, rownames(pbmc@scale.data))
genes.use <- intersect(genes.use, rownames(stim@scale.data))
```


# Canonical Correlation Analysis (CCA)

```{r}
# Run CCA to identify common sources of variation between the two datasets

# CCA finds linear combinations of features across the conditions that are maximally           correlated.  In short, it searches for patterns in the datasets that are common.

immune.combined <- RunCCA(pbmc, stim, genes.use = genes.use, num.cc = 30)

# Visualize CCA results

  # plot CC1 vs CC2
p1 <- DimPlot(object = immune.combined, reduction.use = "cca", group.by = "stim", pt.size = 0.5, do.return = TRUE)
  # violin plot to visualize CC1 and CC2
p2 <- VlnPlot(object = immune.combined, features.plot = "CC1", group.by = "stim", do.return = TRUE, point.size.use = 0.001)
p3 <- VlnPlot(object = immune.combined, features.plot = "CC2", group.by = "stim", do.return = TRUE, point.size.use = 0.001)
plot(p1)
plot_grid(p2, p3)

```

# Choose CCs

```{r}

# Much like selecting the number of PCs to use in the first analysis, we must select the number of CCs to use here.  The MetagenBicorPlot function below is analogous to the elbow plot method used in the first analysis of 1 dataset.  You look for saturation (flatline).

# 2 ways to pick CCs, but we won't run the first because it takes too long to compute for the time we have in class.  Feel free to try it on your own!

# # 1.  MetageneBicorPlot
p4 <- MetageneBicorPlot(immune.combined, grouping.var = "stim", dims.eval = 1:30, display.progress = TRUE)

# 2. Heatmap - use the top 500 cells
  # you need three pages to see all the CCs
  # purple = low expression
  # yellow = high
DimHeatmap(object = immune.combined, reduction.type = "cca", cells.use = 500, dim.use = 1:9,
           do.balanced = TRUE)

DimHeatmap(object = immune.combined, reduction.type = "cca", cells.use = 500, dim.use = 10:18,
           do.balanced = TRUE)

DimHeatmap(object = immune.combined, reduction.type = "cca", cells.use = 500, dim.use = 19:30,
           do.balanced = TRUE)

# How many CCs do you think we should pick?

```

# Align the CCA subspaces and visualize

```{r}
# use the first 20 CCs to get a new dimension reduction that you can visualize and use for clustering

immune.combined <- AlignSubspace(immune.combined, reduction.type = "cca", grouping.var = "stim", 
                    dims.align = 1:20)

# visualize the aligned CCA - how does this compare to before alignment?

p1 <- VlnPlot(object = immune.combined, features.plot = "ACC1", group.by = "stim", do.return = "TRUE", point.size.use = 0.01)
p2 <- VlnPlot(object = immune.combined, features.plot = "ACC2", group.by = "stim", do.return = "TRUE", point.size.use = 0.01)
plot_grid(p1, p2)

# object@dr$reduction.type.aligned

```

# Find clusters using the aligned data

```{r}
# Run t-SNE
immune.combined <- Seurat::RunTSNE(immune.combined, reduction.use = "cca.aligned", dims.use = 1:20, do.fast = T)

# cluster based on t-SNE space
immune.combined <- FindClusters(immune.combined, reduction.type = "cca.aligned", resolution = 0.6, dims.use = 1:20)

# Visualize the dataset alignment and combined clustering results on t-SNE
p1 <- TSNEPlot(immune.combined, do.return = T, pt.size = 0.5, group.by = "stim")
p2 <- TSNEPlot(immune.combined, do.label = T,  do.return = T, pt.size = 0.5)
plot_grid(p1, p2)

```


# Differential gene expression of markers conserved across conditions

```{r}

# Identify conserved cell type markers 
  # Conserved cell type makres are those that identify a particular cell type in both control       pbmc and stimulated pbmc).  The FindConservedmarkers function uses differential gene             expression analysis to do this.

# As an example, let's look for markers that are conserved between control and stimulated cells    in cluster 7, defined in the "ident.1" argument.
cluster7.markers <- FindConservedMarkers(immune.combined, ident.1 = 7, grouping.var = "stim", print.bar = FALSE)
# view the first few markers
cluster7.markers[1:9,]

# Confirm that those markers for cluster 7 identified fall into that location on t-SNE.
FeaturePlot(object = immune.combined, features.plot = rownames(cluster7.markers)[1:9], min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)

# Based on those results, what cell type is cluster 7?


```



```{r}
# DO NOT RUN THIS MORE THAN ONCE

# Explore markers of different immune cell populations to try to identify what cell type each cluster contains.  Select candidates based off of prior knowledge of immune cell markers.
FeaturePlot(object = immune.combined, features.plot = c("CD3D", "SELL", "CREM", "CD8A", "GNLY", "CD79A", "FCGR3A", "CCL2", "PPBP"), min.cutoff = "q9", cols.use = c("lightgrey", "blue"), pt.size = 0.5)

# Now, you can rename your clusters as these cell names.  Trust that they are correct :)
new.ident <- c("CD14 Mono", "CD4 Naive T", "CD4 Memory T", "B", "CD16 Mono",
    "T activated", "CD8 T", "NK", "DC", "B activated", "Mk", "pDC", "Eryth")
for (i in 0:12) {
    immune.combined <- RenameIdent(object = immune.combined, old.ident.name = i,
        new.ident.name = new.ident[i + 1])
}

# Plot t-SNE to see your new cluster names
TSNEPlot(immune.combined, do.label = T, pt.size = 0.5)

use SPlitDotPlotGG to plot markers of each cell type (using prior knowledge again).
immune.combined@ident <- factor(immune.combined@ident, levels = (c("pDC", "Eryth",
    "Mk", "DC", "CD14 Mono", "CD16 Mono", "B activated", "B", "CD8 T", "NK",
    "T activated", "CD4 Naive T", "CD4 Memory T")))
markers.to.plot <- c("CD3D", "CREM", "HSPH1", "SELL", "GIMAP5", "CACYBP", "GNLY",
    "NKG7", "CCL5", "CD8A", "MS4A1", "CD79A", "MIR155HG", "NME1", "FCGR3A",
    "VMO1", "CCL2", "S100A9", "HLA-DQA1", "GPR183", "PPBP", "GNG11", "HBA2",
    "HBB", "TSPAN13", "IL3RA", "IGJ")
sdp <- SplitDotPlotGG(immune.combined, genes.plot = rev(markers.to.plot), cols.use = c("blue", 
    "red"), x.lab.rot = T, plot.legend = T, dot.scale = 3, do.return = T, grouping.var = "stim")
````


# Differential gene expression of markers that change across conditions (ctrl vs stimulated).

```{r}
# It's great that we can identify conserved markers across datasets because it allows us to be confident that we have identified subpopulations of cell types that are present in both datasets, ctrl vs stimulated.  However, what is really interesting is what CHANGES when you stimulate the cells.  Let's find out.

# First we need to define some plotting functions to make labeling easier.  Don't worry about trying to understand what each line is doing, it is not interesting.  Just run it!

LabelPoint <- function(plot, genes, exp.mat, adj.x.t = 0, adj.y.t = 0, adj.x.s = 0, 
                       adj.y.s = 0, text.size = 2.5, segment.size = 0.1) {
  for (i in genes) {
    x1 <- exp.mat[i, 1]
    y1 <- exp.mat[i, 2]
    plot <- plot + annotate("text", x = x1 + adj.x.t, y = y1 + adj.y.t, 
                            label = i, size = text.size)
    plot <- plot + annotate("segment", x = x1 + adj.x.s, xend = x1, y = y1 + 
                              adj.y.s, yend = y1, size = segment.size)
  }
  return(plot)
}

LabelUR <- function(plot, genes, exp.mat, adj.u.t = 0.1, adj.r.t = 0.15, adj.u.s = 0.05, 
                    adj.r.s = 0.05, ...) {
  return(LabelPoint(plot, genes, exp.mat, adj.y.t = adj.u.t, adj.x.t = adj.r.t, 
                    adj.y.s = adj.u.s, adj.x.s = adj.r.s, ...))
}

LabelUL <- function(plot, genes, exp.mat, adj.u.t = 0.1, adj.l.t = 0.15, adj.u.s = 0.05, 
                    adj.l.s = 0.05, ...) {
  return(LabelPoint(plot, genes, exp.mat, adj.y.t = adj.u.t, adj.x.t = -adj.l.t, 
                    adj.y.s = adj.u.s, adj.x.s = -adj.l.s, ...))
}

```

```{r}
# Now, let's see how naive T cells respond to stimulation with interferon gamma.

t.cells <- SubsetData(immune.combined, ident.use = "CD4 Naive T", subset.raw = T)
t.cells <- SetAllIdent(t.cells, id = "stim")
avg.t.cells <- log1p(AverageExpression(t.cells, show.progress = FALSE))
avg.t.cells$gene <- rownames(avg.t.cells)

p1 <- ggplot(avg.t.cells, aes(CTRL, STIM)) + geom_point() + ggtitle("CD4 Naive T Cells")
plot(p1)
```

```{r}
# Let's also see how CD14 monocytes respond to stimulation with interferon gamma.

cd14.mono <- SubsetData(immune.combined, ident.use = "CD14 Mono", subset.raw = T)
cd14.mono <- SetAllIdent(cd14.mono, id = "stim")
avg.cd14.mono <- log1p(AverageExpression(cd14.mono, show.progress = FALSE))
avg.cd14.mono$gene <- rownames(avg.cd14.mono)

p2 <- ggplot(avg.cd14.mono, aes(CTRL, STIM)) + geom_point() + ggtitle("CD14 Monocytes")

plot(p2)
```

```{r}
# If you were to go in and find the genes that lie above the linear correlation, you would find these:

genes.to.label1 = c("ISG15", "LY6E", "IFI6", "ISG20", "MX1")
genes.to.label2 = c("IFIT2", "IFIT1")
genes.to.label3 = c("CXCL10", "CCL8")

# Now, plot them on the scatterplots and see where they fall.  These end up being the genes that respond specifically to interferon gamma stimulation, both in naive T cells and CD14 monocytes - an interferon gamma response pathway that many cells have.

p1 <- LabelUR(p1, genes = c(genes.to.label1, genes.to.label2), avg.t.cells, 
    adj.u.t = 0.3, adj.u.s = 0.23)
p1 <- LabelUL(p1, genes = genes.to.label3, avg.t.cells, adj.u.t = 0.5, adj.u.s = 0.4, 
    adj.l.t = 0.25, adj.l.s = 0.25)
p2 <- LabelUR(p2, genes = c(genes.to.label1, genes.to.label3), avg.cd14.mono, 
    adj.u.t = 0.3, adj.u.s = 0.23)
p2 <- LabelUL(p2, genes = genes.to.label2, avg.cd14.mono, adj.u.t = 0.5, adj.u.s = 0.4, 
    adj.l.t = 0.25, adj.l.s = 0.25)
plot_grid(p1,p2)
```

```{r}
# Now, look for responses specific to particular cell types.

# code to grab particular cell types
immune.combined@meta.data$celltype.stim <- paste0(immune.combined@ident, "_", 
    immune.combined@meta.data$stim)
immune.combined <- StashIdent(immune.combined, save.name = "celltype")
immune.combined <- SetAllIdent(immune.combined, id = "celltype.stim")

# CD14 monocytes
CD14mono.interferon.response <- FindMarkers(immune.combined, ident.1 = "CD14 Mono_STIM", ident.2 = "CD14 Mono_CTRL", print.bar = FALSE)
head(CD14mono.interferon.response, 15)

# B cells
b.interferon.response <- FindMarkers(immune.combined, ident.1 = "B_STIM", ident.2 = "B_CTRL", 
    print.bar = FALSE)
head(b.interferon.response, 15)

# CD4 naive T cells
CD4naiveT.interferon.response <- FindMarkers(immune.combined, ident.1 = "CD4 Naive T_STIM", ident.2 = "CD4 Naive T_CTRL", print.bar = FALSE)
head(CD4naiveT.interferon.response, 15)
```


```{r}
# Now visualize the top 8 fold changes for these 3 groups of cells.  Are these unique responses?  How would we find cell-type specific responses?

# FeatureHeatmap

# CD14 monocytes
FeatureHeatmap(immune.combined, features.plot = rownames(CD14mono.interferon.response)[1:8], group.by = "stim", pt.size = 0.25, key.position = "top", max.exp = 3)

# CD4 naive T cells
FeatureHeatmap(immune.combined, features.plot = rownames(CD4naiveT.interferon.response)[1:8], group.by = "stim", pt.size = 0.25, key.position = "top", max.exp = 3)

# B cells
FeatureHeatmap(immune.combined, features.plot = rownames(b.interferon.response)[1:8], group.by = "stim", pt.size = 0.25, key.position = "top", max.exp = 3)

```

# Save the objects

```{r}
# save this work and come back to it later

saveRDS(pbmc, file = paste(directory, "/pbmc_combined.rds", sep = ""))
saveRDS(stim, file = paste(directory, "/stim_combined.rds", sep = ""))
saveRDS(immune.combined, file = paste(directory, "/immune_combined.rds", sep = ""))
```


